README.md
Mailjet Transport for Symfony Mailer Component
==============================================

[![Packagist](https://img.shields.io/packagist/v/sylvaindeloux/symfony-mailjet-transport.svg)](https://packagist.org/packages/sylvaindeloux/symfony-mailjet-transport)
[![Packagist](https://img.shields.io/packagist/dt/sylvaindeloux/symfony-mailjet-transport.svg)](https://packagist.org/packages/sylvaindeloux/symfony-mailjet-transport)
[![Travis build](https://img.shields.io/travis/sylvaindeloux/symfony-mailjet-transport.svg)](https://travis-ci.org/github/sylvaindeloux/symfony-mailjet-transport)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/sylvaindeloux/symfony-mailjet-transport/blob/master/LICENSE.md)

Install the bundle:

    composer require sylvaindeloux/symfony-mailjet-transport

Add it to `config/bundles.php`:

    <?php

    return [
        // ...
        SylvainDeloux\MailjetTransport\MailjetTransportBundle::class => ['all' => true],
    ];

Now you can use your Mailjet account with Symfony Mailer. You just need to configure the `MAILER_DSN` environment variable with your credentials:

* SMTP: `mailjet+smtp://<your api key>:<your api secret>@in-v3.mailjet.com`
* API: `mailjet+api://<your api key>:<your api secret>@api.mailjet.com?version=3.1`

Use Mailjet templates with variables
------------------------------------

If you want to use a custom template instead of a Twig HTML / text body, and inject your own variables:

    $email = (new \SylvainDeloux\MailjetTransport\Mailer\Email())
        // ...
        ->setTemplateId(<your template id>)
        ->setErrorReportingEmail(<your email address for debugging>) // optional, to get a detailled message if template error occurs
        ->setTemplateErrorDeliver() // optional, if you want the mail to be delivered if template error occurs
        ->setVariables(array(
            'key' => 'value',
        ))
    ;


CHANGELOG.md
CHANGELOG
=========

0.5
---

* Cleanup + tests

0.4
---

* Better error handler

0.3
---

* Added properties to Email

0.2
---

* Added Mailjet custom template usage

0.1
---

* First version


tests/Transport/MailjetApiTransportTest.php
<?php

namespace SylvainDeloux\MailjetTransport\Transport;

use PHPUnit\Framework\TestCase;
use Psr\EventDispatcher\EventDispatcherInterface;
use Psr\Log\LoggerInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class MailjetApiTransportTest extends TestCase
{
    /**
     * @dataProvider getTransportData
     */
    public function testToString(MailjetApiTransport $transport, string $expected)
    {
        $this->assertSame($expected, (string) $transport);
    }

    public function getTransportData()
    {
        $client = $this->getMockBuilder(HttpClientInterface::class)->disableOriginalConstructor()->getMock();
        $dispatcher = $this->getMockBuilder(EventDispatcherInterface::class)->disableOriginalConstructor()->getMock();
        $logger = $this->getMockBuilder(LoggerInterface::class)->disableOriginalConstructor()->getMock();

        return [
            [
                (new MailjetApiTransport($client))->setHost('host'),
                'mailjet+api://host?version=3.1',
            ],
            [
                (new MailjetApiTransport($client))->setHost('host')->setVersion('12'),
                'mailjet+api://host?version=12',
            ],
        ];
    }

    // @todo: more tests
}


tests/Transport/MailjetTransportFactoryTest.php
<?php

namespace SylvainDeloux\MailjetTransport\Transport;

use Symfony\Component\Mailer\Test\TransportFactoryTestCase;
use Symfony\Component\Mailer\Transport\Dsn;
use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
use Symfony\Component\Mailer\Transport\TransportFactoryInterface;

class MailjetTransportFactoryTest extends TransportFactoryTestCase
{
    public function getFactory(): TransportFactoryInterface
    {
        return new MailjetTransportFactory($this->getDispatcher(), $this->getClient(), $this->getLogger());
    }

    public function supportsProvider(): iterable
    {
        yield [
            new Dsn('mailjet', 'default'),
            true,
        ];

        yield [
            new Dsn('mailjet+api', 'default'),
            true,
        ];

        yield [
            new Dsn('mailjet+smtp', 'default'),
            true,
        ];
    }

    public function createProvider(): iterable
    {
        $client = $this->getClient();
        $dispatcher = $this->getDispatcher();
        $logger = $this->getLogger();

        $apiTransport = new MailjetApiTransport($client, $dispatcher, $logger);
        $apiTransport->setApiKey(self::USER);
        $apiTransport->setSecretKey(self::PASSWORD);
        $apiTransport->setHost('default');
        $apiTransport->setVersion('3.1');

        $smtpTransport = new EsmtpTransport('default', 587, false, $dispatcher, $logger);
        $smtpTransport->setUsername(self::USER);
        $smtpTransport->setPassword(self::PASSWORD);

        yield [
            new Dsn('mailjet+api', 'default', self::USER, self::PASSWORD),
            $apiTransport,
        ];

        yield [
            new Dsn('mailjet+smtp', 'default', self::USER, self::PASSWORD),
            $smtpTransport,
        ];
    }

    public function unsupportedSchemeProvider(): iterable
    {
        yield [
            new Dsn('mailjet+foo', 'default', self::USER, self::PASSWORD),
            'The "mailjet+foo" scheme is not supported; supported schemes for mailer "mailjet" are: "mailjet", "mailjet+api", "mailjet+smtp".',
        ];
    }

    public function incompleteDsnProvider(): iterable
    {
        yield [new Dsn('mailjet+api', 'default', self::USER)];

        yield [new Dsn('mailjet+api', 'default', null, self::PASSWORD)];
    }
}


src/Mailer/Email.php
<?php

namespace SylvainDeloux\MailjetTransport\Mailer;

use Symfony\Component\Mime\Email as BaseEmail;
use Symfony\Component\Mime\Exception\LogicException;
use Symfony\Component\Mime\Header\Headers;
use Symfony\Component\Mime\Part\AbstractPart;

class Email extends BaseEmail
{
    protected $templateId = null;
    protected $variables = array();
    protected $errorReportingEmail = null;
    protected $templateErrorDeliver = false;
    protected $additionalProperties = array();

    public function __construct(Headers $headers = null, AbstractPart $body = null)
    {
        parent::__construct($headers, $body);

        $this->html('');
        $this->text('');
    }

    public function getTemplateId()
    {
        return $this->templateId;
    }

    public function setTemplateId(int $templateId): self
    {
        $this->templateId = $templateId;

        return $this;
    }

    public function getVariables(): array
    {
        return $this->variables;
    }

    public function setVariables(array $variables): self
    {
        $this->variables = $variables;

        return $this;
    }

    public function setVariable(string $key, string $value): self
    {
        $this->variables[$key] = $value;

        return $this;
    }

    public function getErrorReportingEmail()
    {
        return $this->errorReportingEmail;
    }

    public function setErrorReportingEmail(string $errorReportingEmail): self
    {
        $this->errorReportingEmail = $errorReportingEmail;

        return $this;
    }

    public function isTemplateErrorDeliver(): bool
    {
        return $this->templateErrorDeliver;
    }

    public function setTemplateErrorDeliver(bool $templateErrorDeliver = true): self
    {
        $this->templateErrorDeliver = $templateErrorDeliver;

        return $this;
    }

    public function getAdditionalProperties(): array
    {
        return $this->additionalProperties;
    }

    public function setAdditionalProperties(array $additionalProperties): self
    {
        $this->additionalProperties = $additionalProperties;

        return $this;
    }

    public function addProperty(string $key, string $value): self
    {
        $this->additionalProperties[$key] = $value;

        return $this;
    }

    public function ensureValidity()
    {
        if (null === $this->templateId) {
            if (count($this->variables)) {
                throw new \LogicException('A template id is required.');
            }
        }

        parent::ensureValidity();
    }

    /**
     * @internal
     */
    public function __serialize(): array
    {
        return [$this->templateId, $this->variables, $this->errorReportingEmail, $this->templateErrorDeliver, $this->additionalProperties, parent::__serialize()];
    }

    /**
     * @internal
     */
    public function __unserialize(array $data): void
    {
        [$this->templateId, $this->variables, $this->errorReportingEmail, $this->templateErrorDeliver, $this->additionalProperties, $parentData] = $data;

        parent::__unserialize($parentData);
    }
}


src/Transport/MailjetApiTransport.php
<?php

namespace SylvainDeloux\MailjetTransport\Transport;

use SylvainDeloux\MailjetTransport\Mailer\Email as MailjetEmail;
use Symfony\Component\Mailer\Envelope;
use Symfony\Component\Mailer\Exception\HttpTransportException;
use Symfony\Component\Mailer\SentMessage;
use Symfony\Component\Mailer\Transport\AbstractApiTransport;
use Symfony\Component\Mime\Email;
use Symfony\Contracts\HttpClient\ResponseInterface;

class MailjetApiTransport extends AbstractApiTransport
{
    private $apiKey;
    private $secretKey;
    private $version = '3.1';

    public function __toString(): string
    {
        return sprintf('mailjet+api://%s?version=%s', $this->host, $this->version);
    }

    protected function doSendApi(SentMessage $sentMessage, Email $email, Envelope $envelope): ResponseInterface
    {
        $response = $this->client->request('POST', $this->getEndpoint().'/send', array(
            'json' => $this->getPayload($email, $envelope),
            'auth_basic' => array($this->apiKey, $this->secretKey),
        ));

        $result = $response->toArray(false);

        if (array_key_exists('ErrorMessage', $result)) {
            throw new HttpTransportException(sprintf('Unable to send an email: %s (error code: %s).', $result['ErrorMessage'], $result['ErrorCode']), $response);
        }

        $mailjetMessage = reset($result['Messages']);

        if ('error' === $mailjetMessage['Status']) {
            throw new HttpTransportException(sprintf('Unable to send an email: %s (error code: %s, related to "%s").', $mailjetMessage['Errors'][0]['ErrorMessage'], $mailjetMessage['Errors'][0]['ErrorCode'], implode('", "', $mailjetMessage['Errors'][0]['ErrorRelatedTo'])), $response);
        }

        if ('success' === $mailjetMessage['Status']) {
            $sentMessage->setMessageId($mailjetMessage['To'][0]['MessageUUID']);
        }

        return $response;
    }

    private function getPayload(Email $email, Envelope $envelope): array
    {
        $payload = array(
            'Subject' => $email->getSubject(),
            'From' => array(
                'Email' => $envelope->getSender()->getAddress(),
            ),
            'To' => array(),
            'Attachments' => array(),
            'InlinedAttachments' => array(),
        );

        if ($email->getTextBody()) {
            $payload['TextPart'] = $email->getTextBody();
        }

        if ($email->getHtmlBody()) {
            $payload['HTMLPart'] = $email->getHtmlBody();
        }

        if ('' !== $envelope->getSender()->getName()) {
            $payload['From']['Name'] = $envelope->getSender()->getName();
        } elseif ('' !== $email->getFrom()[0]->getName()) {
            $payload['From']['Name'] = $email->getFrom()[0]->getName();
        }

        $recipients = $this->getRecipients($email, $envelope);
        foreach ($recipients as $recipient) {
            $payload['To'][] = array(
                'Email' => $recipient->getAddress(),
                'Name' => $recipient->getName(),
            );
        }

        $attachments = $email->getAttachments();
        foreach ($attachments as $attachment) {
            $headers = $attachment->getPreparedHeaders();
            $disposition = $headers->getHeaderBody('Content-Disposition');

            $att = array(
                'Base64Content' => $attachment->bodyToString(),
                'ContentType' => $headers->get('Content-Type')->getBody(),
            );

            if ($name = $headers->getHeaderParameter('Content-Disposition', 'name')) {
                $att['Filename'] = $name;
            }

            if ('inline' === $disposition) {
                $payload['InlinedAttachments'][] = $att;
            } else {
                $payload['Attachments'][] = $att;
            }
        }

        if ($email instanceof MailjetEmail) {
            if ($email->getTemplateId()) {
                $payload['TemplateLanguage'] = true;
                $payload['TemplateID'] = $email->getTemplateId();
            }

            if (count($email->getVariables())) {
                $payload['Variables'] = $email->getVariables();
            }

            if ($email->getErrorReportingEmail()) {
                $payload['TemplateErrorReporting'] = array(
                    'Email' => $email->getErrorReportingEmail(),
                );
            }

            if ($email->isTemplateErrorDeliver()) {
                $payload['TemplateErrorDeliver'] = true;
            }

            if (count($email->getAdditionalProperties())) {
                $payload = array_merge($payload, $email->getAdditionalProperties());
            }
        }

        return array('Messages' => array($payload));
    }

    protected function getEndpoint(): string
    {
        return sprintf('https://%s/v%s', $this->host, $this->version);
    }

    public function setApiKey(string $apiKey): self
    {
        $this->apiKey = $apiKey;

        return $this;
    }

    public function setSecretKey(string $secretKey): self
    {
        $this->secretKey = $secretKey;

        return $this;
    }

    public function setVersion(?string $version): self
    {
        $this->version = $version;

        return $this;
    }
}


src/Transport/MailjetTransportFactory.php
<?php

namespace SylvainDeloux\MailjetTransport\Transport;

use Symfony\Component\Mailer\Exception\UnsupportedSchemeException;
use Symfony\Component\Mailer\Transport\AbstractTransportFactory;
use Symfony\Component\Mailer\Transport\Dsn;
use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
use Symfony\Component\Mailer\Transport\TransportInterface;

final class MailjetTransportFactory extends AbstractTransportFactory
{
    public function create(Dsn $dsn): TransportInterface
    {
        $scheme = $dsn->getScheme();
        $user = $this->getUser($dsn);
        $password = $this->getPassword($dsn);
        $host = $dsn->getHost();
        $port = $dsn->getPort(587);

        if ('mailjet+api' === $scheme) {
            $transport = new MailjetApiTransport($this->client, $this->dispatcher, $this->logger);
            $transport->setApiKey($user);
            $transport->setSecretKey($password);
            $transport->setHost($host);
            $transport->setVersion($dsn->getOption('version', '3.1'));
            return $transport;
        }

        if ('mailjet+smtp' === $scheme) {
            $transport = new EsmtpTransport($host, $port, false, $this->dispatcher, $this->logger);
            $transport->setUsername($user);
            $transport->setPassword($password);
            return $transport;
        }

        throw new UnsupportedSchemeException($dsn, 'mailjet', $this->getSupportedSchemes());
    }

    protected function getSupportedSchemes(): array
    {
        return array('mailjet', 'mailjet+api', 'mailjet+smtp');
    }
}
